File: cpp/05_smallest_multiple.cpp
#include <iostream>

int gcd(int a, int b){
    if (a % b == 0) return b;
    return gcd(b, a % b);
}

int lcm(int a, int b){
    // prevent integer overflow by performing division first
    return a / gcd(a, b) * b;
}

int smallest_multiple(int n){
    int mult = 2520;
    for (int i = 11; i <= n; i++){
        mult = lcm(mult, i);
    }
    return mult;
}

int main() {
    std::cout << smallest_multiple(20) << "\n";
}

File: c#/_05_SmallestMultiple.cs
using System;

public class _05_SmallestMultiple{
    public static int Euler05(){
        int res = 1;
        for (int i = 2; i <= 20; i++){
            res = res / GCD(res, i) * i;
        }
        return res;
    }
    
    private static int GCD(int a, int b){
        while (b > 0){
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    
    public static void Main(string[] args){
        Console.WriteLine(Euler05());
    }
}
